pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -571,6 +575,17 @@ protected void handleSend(CommandSend send, ByteBuf headersAndPayload) {
             printSendCommandDebug(send, headersAndPayload);
         }
 
+        // avoid processing non-persist message if reached max concurrent-message limit
+        if (producer.isNonPersistentTopic() && !nonPersistentMessageSemaphore.tryAcquire()) {
+            final long producerId = send.getProducerId();
+            final long sequenceId = send.getSequenceId();
+            service.getTopicOrderedExecutor().submitOrdered(producer.getTopic(), SafeRun.safeRun(() -> {
```

Comment:
Why schedule it in ordered thread? We are in the io thread, we should write it already from this thread.